[Core, iOS, UWP, Android] Popup Control Implementation

ahoefling opened the pull request:
### Description of Change ###
This is an incomplete PR as the iOS implementation is not complete. I am submitting it as I have been talking with Xamarin Team members and Community members for almost an entire year about this Popup Control. The goal is to submit it now and start getting feedback and help on the iOS implementation

Implementation of Popup Control (#1778), this change adds native popup control for the major platforms Android, UWP, iOS (incomplete) as defined in the initial work item
- Android - `Dialog`
- UWP - `Flyout`
- iOS - `Popover`

I have been dog fooding the android implementation as the official Popup Control for our internal projects at FileOnQ. The API has worked very well for us over the past 8+ months. 

### Issues Resolved ### 
- fixes #1778

### API Changes (INavigation) ###
Added ability to show Popups from the `INavigation` interface

Added:
 - `Task<T> ShowPopup<T>(Popup<T> popup);`

### API Changes (BasePopup) ###
Added `BasePopup` control designed for rendering a popup that does not care about the return result

Added:
- `View View { get; set; }`
- `Color Color { get; set; }`
- `Color BorderColor { get; set; } // UWP ONLY - wasn't originally in spec`
- `View Anchor { get; set; }`
- `Size Size { get; set; }`
- `bool IsLightDismissEnabled { get; set; }`
- `event EventHandler<PopupDismissedEventArgs> Dismissed;`

 ### API Changes (Popup<T>) ###
Added `Popup<T>` control designed for rendering a popup and returning a value

Added:
- `View View { get; set; }`
- `Color Color { get; set; }`
- `Color BorderColor { get; set; } // UWP ONLY - wasn't originally in spec`
- `View Anchor { get; set; }`
- `Size Size { get; set; }`
- `bool IsLightDismissEnabled { get; set; }`
- `event EventHandler<PopupDismissedEventArgs> Dismissed;`
- `void Reset()`
- `void Dismiss(T result)`
- `Task<T> Result`

### Library Addition ###
In one of the new Control Gallery pages I needed device window information and it was easiest to pull in [Xamarin.Essentials](https://github.com/xamarin/Essentials) for that test project. 

### Platforms Affected ### 
<!-- Please list all platforms affected by these changes -->

- Core/XAML (all platforms)
- iOS
- Android
- UWP

### Usage ###
The popup control can be used just like a `ContentPage` or `ContentView`, the power in our implementation is being able to create a simple XAML page that defines how your popup is rendered.

**SimplePopup.xaml**
Code snippet in ControlGallery
```xaml
<?xml version="1.0" encoding="utf-8" ?>
<Popup xmlns="http://xamarin.com/schemas/2014/forms"
       xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
       xmlns:d="http://xamarin.com/schemas/2014/forms/design"
       xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
       mc:Ignorable="d"
       x:Class="Xamarin.Forms.Controls.SimplePopup">
    <Popup.View>
        <StackLayout VerticalOptions="FillAndExpand">
            <Label Text="Custom Dialog Popup"
                   FontSize="20"
                   FontAttributes="Bold"
                   HorizontalTextAlignment="Center"/>
            <Label Text="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip "
                   VerticalOptions="CenterAndExpand" 
                   HorizontalOptions="CenterAndExpand"
                   Margin="10, 0"/>
            <Button Text="OKAY"
                    VerticalOptions="EndAndExpand"
                    Clicked="Button_Clicked" />
        </StackLayout>
    </Popup.View>
</Popup>
```

### Behavioral/Visual Changes ###
not applicable as this is a new API

### Before/After Screenshots ###
UWP and Android gifs that demonstrate basic positioning functionality

![dialog-window-posiiton-uwp](https://user-images.githubusercontent.com/17751436/74692646-2aab7980-51b6-11ea-8090-bd39a20c1469.gif)
![dialog-window-posiiton-android](https://user-images.githubusercontent.com/17751436/74692654-2f702d80-51b6-11ea-880b-e56ef682a994.gif)

### Testing Procedure ###
I have created a series of tests in the Control Gallery App, there is much more testing that needs to be done and samples that should be created prior to merging this change.

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### Items Remaining Checklist
This checklist provides everyone with an overall status of where the PR currently is

- [x] Android implementation
- [x] UWP implementation
- [x] iOS implementation
- [x] Core implementation
- [ ] Example code

VladislavAntonyuk commented on Xamarin.Forms.ControlGallery.WPF/Xamarin.Forms.ControlGallery.WPF.csproj:
```diff
@@ -56,6 +56,9 @@
     <PackageReference Include="Microsoft.Maps.MapControl.WPF" Version="1.0.0.3" />
     <PackageReference Include="OpenTK" Version="3.0.1" />
     <PackageReference Include="OpenTK.GLControl" Version="3.0.1" />
+    <PackageReference Include="Xamarin.Essentials">
```
WPF is not supported. Do we need it?

ahoefling replied:
There is currently no WPF support this is not needed, I can pull this out.

ahoefling replied:
This was resolved when I rebased the branch on 4.6.0

VladislavAntonyuk commented on Xamarin.Forms.Core/Popup/Popup.cs:
```diff
@@ -0,0 +1,138 @@
+		}
+	}
+
+	public abstract class Popup<T> : BasePopup
```
1 class in file

ahoefling replied:
Thanks for the feedback, I really should add a `BasePopup` and `Popup` class file.

ahoefling replied:
This has been removed during the rebase process

ahoefling replied:
Added new class files in latest changes

GalaxiaGuy commented:
I'm curious about how this is considered vs https://github.com/xamarin/Xamarin.Forms/pull/8551 (Modal Page with transparent background).

Their implementation is very different, but there is quite a lot of overlap conceptually.

ahoefling commented:
@GalaxiaGuy 
>I'm curious about how this is considered vs #8551 (Modal Page with transparent background).
>
>Their implementation is very different, but there is quite a lot of overlap conceptually.

I believe these 2 APIs can live in harmony and compliment each other. AFAIK the Modal API that @jfversluis is working on is designed to have a transparent background for the `ContentPage` and simulate popup features. A very useful API to solve rendering problems that may not be available by a normal `ContentPage` or this Popup Control API. Please correct me if I have a misunderstanding.

The goal of this new API is to render 100% native popups. Instead of rendering a stack of `ContentPages` it will render the correct native control and render the `VisualElement` inside of that native popup control. A major advantage that this API has over the Modal API is it does not have the Anchoring feature. You can place a Popup relative to another control on the screen. This can be useful for quick tooltips and other advanced popup techniques.

I do not want to take away from any work that has been done on the other PR, but I do believe these 2 APIs can live together as I see them having different use-cases.

jsuarezruiz commented on Xamarin.Forms.Platform.iOS/PopupWrapper.cs:
```diff
@@ -0,0 +1,85 @@
+
+			if (!popup.Size.IsZero)
+			{
+				PreferredContentSize = new CGSize(popup.Size.Width, popup.Size.Height);
```
Testing I always have the same popup size:
![ios-popup-size](https://user-images.githubusercontent.com/6755973/78222516-aee76100-74c5-11ea-8ea9-9f031490c6d9.gif)

ahoefling replied:
I'm going to mark this conversation as resolved since the iOS implementation is further along and is mostly working

ahoefling commented:
@jsuarezruiz I have rebased the branch on 4.6.0, I am going to start reviewing the comments and suggestions.

As of right now the iOS implementation is not complete and is only partially completed. As I get time, I am going to finish up the iOS implementation. Big first step here is getting the branch rebased.

ahoefling commented on Xamarin.Forms.Core/Popup/Popup.cs:
```diff
@@ -0,0 +1,138 @@
+		/// This color sets the native background color of the <see cref="Popup"/>, which is
+		/// independent of any background color configured in the actual View.
+		/// </remarks>
+		public Color Color { get; set; }
```
None of these properties were made bindable because you have 100% control of the `View` property where you would apply your `BindingContext`. I can see some good arguments for bindable properties on this class. I am going to update the following properties to be bindable
* **Color**
* **BorderColor**
* **Size**

ahoefling commented on Xamarin.Forms.Core/Popup/Popup.cs:
```diff
@@ -0,0 +1,138 @@
+		/// This color sets the native border color of the <see cref="Popup"/>, which is
+		/// independent of any border color configured in the actual view.
+		/// </remarks>
+		public Color BorderColor { get; set; } // UWP ONLY - wasn't originally in spec
```
What is the best way to implement this in core libraries? Is it as simple as placing a pre-proccessor directive block around the property? If not do you have any examples I can look at so I follow the Xamarin.Forms standard

ahoefling replied:
I'm going to need to add another platform specific property for iOS (popover arrow), do you have any recommendations or controls I can look at as an example for this?

Is it as simple as just creating a pre-processor directive such as `#if WINDOWS_UWP`

ahoefling commented:
My latest push to the PR is a partial implementation for iOS and a bunch of restructuring of the code. It is missing many key features such as layout.

@jsuarezruiz feel free to take a look and leave comments

myroot commented on Xamarin.Forms.Core/Popup/BasePopup.cs:
```diff
@@ -0,0 +1,97 @@
+
+namespace Xamarin.Forms
+{
+	public abstract class BasePopup : View
```
Do you have any reason inherit from `View`? I think it should be inherit `Element` instead `View` because it is not `View`, it can't layouting by `Layout` class

ahoefling replied:
The `BasePopup` needs to inherit from `View` because it has layout context which allows you to place the `Popup` in 9 positions on the screen
* top left
* top center
* top right
* right middle
* bottom right
* bottom center
* bottom left
* left middle
* center

The `View` object that is encapsulated in the `BasePopup` is used for rendering the content that is rendered inside of the native popup control. This content also has layout options within the context of the native popup which is why we require the Layout Options for both the native popup and the content that is rendered on the inside of the popup.

**When I work on my next push for this PR I am going to double check all of this to make sure this is how it is working.** If I can update this to `Element` I'll make the change. Thanks for the feedback 👍

myroot replied:
hm, if BasePopup was inherit from View, It can be contained by Layout class, but it seems not allow it, 
 for eaxmple, it is progrmatically allowed code, but actually, it is not working
``` c#
  new StackLayout {
       Children = {
              new Popup {
                   View = new Label { ... }
              }
  }
```

ahoefling replied:
I see your point now, thanks for clarifying. The Popup should not be used like that and if it is inherited from `View`, it will compile to be used like that. I agree with you 100% now and it doesn't make sense for `BasePopup` to inherit from `View`. I will take another look at this and use `Element` or similar parent class on my next set of work for this PR.

ahoefling replied:
I have updated the `BasePopup` to inherit from `VisualElement`

AlleSchonWeg commented:
There is already a nice popup control: https://github.com/rotorgames/Rg.Plugins.Popup
Perhaps you could integrate it into xamarin.

jsuarezruiz commented:
@ahoefling Sorry, could you rebase master to fix the conflicts?. Thanks!.

ahoefling commented:
@AlleSchonWeg 
> There is already a nice popup control: https://github.com/rotorgames/Rg.Plugins.Popup
Perhaps you could integrate it into xamarin.

This is an implementation of the spec written by the maintainers (Xamarin.Forms Team). When I submit a PR I try to follow as close as possible to the original spec which has been approved by the team.

@jsuarezruiz 
> Sorry, could you rebase master to fix the conflicts?. Thanks!.

No problem, I'll take a look at rebasing this later today. This has been a long PR process and it is expected to need a couple rebases

ahoefling commented:
These rebases are getting easier 😁😁

I have rebased this PR on master, but have not added anything new. I haven't had time to get back to the needed iOS changes. I will take a look when I get some time and if anyone is interested in helping out, or the Xamarin Team wants to finish it up I am more than happy to explain where I left off. This PR is 100% coming from my free time, I only get a chance to work on it as I have time

ahoefling commented:
@jsuarezruiz @myroot 
I have updated rebased the PR on main and added some changes to the PR based on conversations we have been having

- `BasePopup` now inherits from `VisualElement` instead of `View`
- Added `VerticalOptions` and `HorizontalOptions` to `BasePopup` to handle positioning the Popup Control on the screen
- Implemented layout for iOS (see screenshots below)

Centered:
![image](https://user-images.githubusercontent.com/17751436/87498105-0ba9fa00-c625-11ea-8a7b-101fa7775050.png)

Top Left
![image](https://user-images.githubusercontent.com/17751436/87498146-1cf30680-c625-11ea-81ec-d8ea1c257c2a.png)

Right
![image](https://user-images.githubusercontent.com/17751436/87498181-2c724f80-c625-11ea-9cf7-e935415a34df.png)

# What's Left for iOS
- Tapping outside the popup should dismiss the popup control
- UIPopoverControlArrow customization, currently it is hardcoded to point ⬆

ahoefling commented on Xamarin.Forms.Platform.Android/Renderers/PopupRenderer.cs:
```diff
@@ -0,0 +1,244 @@
+namespace Xamarin.Forms.Platform.Android
+{
+	// todo - convert to IVisualElementRenderer since we don't need everything from VisualElementRenderer
+	public class PopupRenderer : VisualElementRenderer<BasePopup>, IDialogInterfaceOnCancelListener
```
This review must have slipped through the cracks, I thought I checked this back in April, guess not 😕. I will take a look at this for my next update and see if I can convert all renderers to use `IVisualElementRenderer` if applicable

ahoefling replied:
See the latest update, I have refactored the Android Renderer to use IVisualElementRenderer and tried to follow the best practices in recent Renderers that followed a similar pattern. I think it meets everything and I need to go through and do the same for iOS and UWP now.

ahoefling commented on Xamarin.Forms.Platform.Android/Renderers/PopupRenderer.cs:
```diff
@@ -0,0 +1,244 @@
+
+			_dialog.Show();
+		}
+
```
I have implemented BindableProperties vis `OnPropertyChanged` for the android PopupRenderer
* Color
* VerticalOptions
* HorizontalOptions
* Size

There is a new BindablePopup in the Popover Gallery that demonstrates how this works

ahoefling commented:
Just pushed another update as I work through some of the finer details in the review. I refactored the **Android** implementation to use `IVisualElementRenderer` as suggested by @jsuarezruiz and implemented BindableProperties for that renderer.  See screenshot below:

![popup_bindable_properties](https://user-images.githubusercontent.com/17751436/87597330-9f300900-c6bf-11ea-9d31-628a77eaa703.gif)


_Expect to see a few more updates over the next couple days as I have some time dedicated to get more work done on this_

ahoefling commented on Xamarin.Forms.Platform.UAP/PopupRenderer.cs:
```diff
@@ -0,0 +1,251 @@
+		void SetLayout()
+		{
+			LightDismissOverlayMode = LightDismissOverlayMode.On;
+			SetDialogPosition(Element.VerticalOptions, Element.HorizontalOptions);
```
I am unable to get the `Flyout.Placement` to update correctly when the `VerticalOptions` or `HorizontalOptions` are updated for UWP. I've tried the following:
- Updating the `_flyoutStyle` and `_panelStyle`
- Hiding the flyout and trying to redraw it
- Nulling out the attached Flyout and re-adding it to the parent control
- Various orders of styles and placement updates

It almost appears that I need to completely dispose of the control and re-initialize it but then we risk losing state. If anyone has suggestions on how we can get this to work I am all ears.

ahoefling commented:
I am closing this PR

I spoke with @PureWeen @jsuarezruiz and @jfversluis today about this control. We have decided it is best to be submitted to the Xamarin Community Toolkit. I have created a new specification over at the community toolkit where you can follow for updates.
Xamarin/XamarinCommunityToolkit#292
